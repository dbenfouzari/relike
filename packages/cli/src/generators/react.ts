import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
