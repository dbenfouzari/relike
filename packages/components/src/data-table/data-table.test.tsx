import { render, screen } from "@testing-library/react";

import Icon from "../icon";
import Icons from "../icons";
import DataTable, { buildDataColumn } from "./data-table";

const data = [{ name: "John Doe", age: 23, gender: "male" }] as const;

/** Defines data type */
type Data = typeof data;

describe("DataTable", () => {
  it("should render successfully", () => {
    const DataColumn = buildDataColumn<Data>();

    render(
      <DataTable data={data}>
        <DataColumn
          value="name"
          label="Name"
          formatValue={(value) => (
            <>
              <Icon icon={Icons.person} />
              <span>{value}</span>
            </>
          )}
        />
        <DataColumn renderCell={({ value }) => <h3>{value}</h3>} value="age" label="Age" />
        <DataColumn value="gender" label="Gender" />
      </DataTable>,
    );

    expect(screen.getByRole("table")).toMatchSnapshot();
  });

  it("buildDataColumn should render nothing", () => {
    const DataColumn = buildDataColumn<Data>();
    const { container } = render(<DataColumn label="Name" value="name" />);

    // Disable ESLint because I want to test that it renders nothing.
    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toBe(null);
  });
});
