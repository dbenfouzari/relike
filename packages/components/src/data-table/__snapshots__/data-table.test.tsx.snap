// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render successfully 1`] = `
<div
  role="table"
>
  <div
    class="header row"
    role="columnheader"
  >
    <div
      class="cell"
      role="cell"
      style="flex: 1;"
    >
      Name
    </div>
    <div
      class="cell"
      role="cell"
      style="flex: 1;"
    >
      Age
    </div>
    <div
      class="cell"
      role="cell"
      style="flex: 1;"
    >
      Gender
    </div>
  </div>
  <div
    class="content"
  >
    <div
      class="row"
      role="row"
    >
      <div
        class="cell"
        role="cell"
      >
        <span
          class="jsx-786785084 icon material-icons"
        >
          person
        </span>
        <span>
          John Doe
        </span>
      </div>
      <h3>
        23
      </h3>
      <div
        class="cell"
        role="cell"
      >
        male
      </div>
    </div>
  </div>
</div>
`;
